src/bam/BamIndexReader.ts

```diff
@@ -0,0 +1,195 @@
+}
+
+const BAI_MAGIC = 21578050;
+const PSEUDO_BIN_MAGIC = 37450;
+
+/**
+ * 
+ * @param bamIndexRefData 
+ * @param refId 
+ * @param start 
+ * @param end 
+ */
+export async function blocksForRange(indexData: BamIndexData, refId: number, 
+        start: number, end: number): Promise<Array<Chunk>> {
+    const overlappingBins: Array<number> = reg2bins(start, end);
+    const binIndex: BinIndex = indexData.refData[refId].binIndex;
+    const linearIndex: Array<VirtualOffset> = indexData.refData[refId].linearIndex;
+    
+    // Get all chunks for overlapping bins.
+    let allChunks: Array<Chunk> = [];
+    for (let bin in binIndex) {
+        if (!overlappingBins.includes(Number(bin))) continue;
+        allChunks = allChunks.concat(binIndex[bin]);
+    }
+
+    // Use the linear index to find minimum file position of chunks that could 
+    // contain alignments in the region.
+    let lowest: VirtualOffset | undefined = undefined;
+    const minLin = Math.min(start >> 14, linearIndex.length - 1);
+    const maxLin = Math.max(end >> 14, linearIndex.length - 1);
+    for (let i = minLin; i <= maxLin; i++) {
+        let offset: VirtualOffset = linearIndex[i];
+        if (offset === undefined) continue;
+        if (lowest === undefined || isVOLessThan(offset, lowest)) {
+            lowest = offset;
+        }
+    }
+
+    return optimizeChunks(allChunks, lowest);
+}
+
+function isVOLessThan(first: VirtualOffset, second: VirtualOffset): boolean {
+    return first.blockPosition < second.blockPosition ||
+            (first.blockPosition === second.blockPosition && first.dataPosition < second.dataPosition);
+}
+
+function optimizeChunks(chunks: Array<Chunk>, lowest?: VirtualOffset): Array<Chunk> {
+    if (chunks.length === 0) return []
+    let mergedChunks: Array<Chunk> = [];
+
+    chunks.sort(function (c0, c1) {
+        var dif = c0.start.blockPosition - c1.start.blockPosition;
+        if (dif != 0) {
+            return dif;
+        } else {
+            return c0.start.dataPosition - c1.start.dataPosition;
+        }
+    });
+
+    let currentMergedChunk: Chunk | undefined = undefined;
+    for (let chunk of chunks) {
+        if (lowest !== undefined && isVOLessThan(chunk.end, lowest)){
+            continue;
+        }
+        if (currentMergedChunk === undefined) {
+            currentMergedChunk = chunk;
+            mergedChunks.push(currentMergedChunk);
+        }
+
+        // Merge chunks that are withing 65k of each other
+        if ((chunk.start.blockPosition - currentMergedChunk.end.blockPosition) < 65000) {
+            if (isVOLessThan(currentMergedChunk.end, chunk.end)) {
+                currentMergedChunk.end = chunk.end;
+            }
+        } else {
+            currentMergedChunk = chunk;
+            mergedChunks.push(currentMergedChunk);
+        }
+    }
+
+    return mergedChunks;
+}
+
+export async function readBamIndex(indexDataLoader: DataLoader): Promise<BamIndexData> {
```

Comment:
This might be better in terms of forward-compatibility to not return the full index, but make this a `BamIndex` class and cache the full index as a field. Then have a getter for `firstAlignmentBlock` and put `blocksForRange` in the class. Then, theoretically, if you come up with a better tree-like index, you aren't relying on everything being in memory.

Though, I'm thinking that if you did have a "cached" version of an index (in a different format), there would be changes elsewhere anyways, so it might not even be worth the changes.

Reply:
I agree with everything you said, but since creating our own index is not currently on the radar, I'd rather not make all the changes. Also if we did invent a a new type of index, we'd probably end up making a new reader too so meh.